File: keybind-common/src/main/java/io/github/solusmods/eternalcore/keybind/api/KeybindingManager.java
package io.github.solusmods.eternalcore.keybind.api;

import dev.architectury.event.events.client.ClientTickEvent;
import dev.architectury.registry.client.keymappings.KeyMappingRegistry;

import java.util.ArrayList;

public class KeybindingManager {
    private static final ArrayList<EternalKeybinding> keybindings = new ArrayList<>();

    public static <T extends EternalKeybinding> void register(T... keybindings) {
        for (final EternalKeybinding keybinding : keybindings) {
            register(keybinding);
        }
    }

    public static <T extends EternalKeybinding> T register(T keybinding) {
        keybindings.add(keybinding);
        KeyMappingRegistry.register(keybinding);
        return keybinding;
    }

    public static void init() {
        // Execute Actions on press
        ClientTickEvent.CLIENT_POST.register(instance -> {
            for (final EternalKeybinding keybinding : keybindings) {
                if (keybinding.isDown()) {
                    keybinding.getAction().onPress();
                } else if (keybinding.getRelease() != null) {
                    keybinding.getRelease().run();
                }
            }
        });
    }
}

File: keybind-common/src/main/kotlin/io/github/solusmods/eternalcore/keybind/api/KeybindingManager.kt
package io.github.solusmods.eternalcore.keybind.api

import dev.architectury.event.events.client.ClientTickEvent
import dev.architectury.registry.client.keymappings.KeyMappingRegistry
import net.minecraft.client.Minecraft

object KeybindingManager {
    private val keybindings = ArrayList<EternalKeybinding>()

    @JvmStatic
    fun <T : EternalKeybinding?> register(vararg keybindings: T?) {
        for (keybinding in keybindings) {
            KeybindingManager.register<EternalKeybinding?>(keybinding)
        }
    }

    fun <T : EternalKeybinding?> register(keybinding: T?): T? {
        keybindings.add(keybinding!!)
        KeyMappingRegistry.register(keybinding)
        return keybinding
    }

    fun init() {
        // Execute Actions on press
        ClientTickEvent.CLIENT_POST.register(ClientTickEvent.Client { instance: Minecraft ->
            for (keybinding in keybindings) {
                if (keybinding.isDown) {
                    keybinding.action!!.onPress()
                } else keybinding.release?.run()
            }
        })
    }
}
